File: crates/daemon-core/src/domain.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

pub type VcsType = codebar_contracts::domain::VcsType;
pub type TrustLevel = codebar_contracts::domain::TrustLevel;
pub type ProviderKind = codebar_contracts::domain::ProviderKind;
pub type WorktreeSource = codebar_contracts::domain::WorktreeSource;
pub type WorktreeLifecycleState = codebar_contracts::domain::WorktreeLifecycleState;
pub type WorktreeCleanupPolicy = codebar_contracts::domain::WorktreeCleanupPolicy;
pub type TaskStatus = codebar_contracts::domain::TaskStatus;
pub type SessionLaunchMode = codebar_contracts::domain::SessionLaunchMode;
pub type SessionState = codebar_contracts::domain::SessionState;
pub type RunLauncherType = codebar_contracts::domain::LauncherType;
pub type RunExitReason = codebar_contracts::domain::RunExitReason;
pub type RunStatus = codebar_contracts::domain::RunAttemptStatus;
pub type PlanMode = codebar_contracts::domain::PlanMode;
pub type PlanStatus = codebar_contracts::domain::PlanStatus;
pub type PlanStepStatus = codebar_contracts::domain::PlanStepStatus;
pub type SkillProfileSource = codebar_contracts::domain::SkillProfileSource;
pub type ApprovalActionType = codebar_contracts::domain::ApprovalActionType;
pub type ApprovalStatus = codebar_contracts::rpc::ApprovalStatus;
pub type EventEntityType = codebar_contracts::events::EventEntityType;
pub type EventSource = codebar_contracts::events::EventSource;

pub type Workspace = codebar_contracts::domain::Workspace;
pub type Worktree = codebar_contracts::domain::Worktree;
pub type Task = codebar_contracts::domain::Task;
pub type Session = codebar_contracts::domain::Session;
pub type RunAttempt = codebar_contracts::domain::RunAttempt;
pub type Plan = codebar_contracts::domain::Plan;
pub type PlanStep = codebar_contracts::domain::PlanStep;
pub type SkillProfile = codebar_contracts::domain::SkillProfile;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RecoveryBinding {
    pub session_id: String,
    pub provider: ProviderKind,
    pub provider_session_id: Option<String>,
    pub worktree_path: Option<String>,
    pub run_attempt_id: Option<String>,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ApprovalRequest {
    pub id: String,
    pub session_id: String,
    pub task_id: String,
    pub action_type: ApprovalActionType,
    pub title: String,
    pub description: String,
    pub payload: HashMap<String, Value>,
    pub status: ApprovalStatus,
    pub created_at: String,
    pub resolved_at: Option<String>,
}

pub type EventEnvelope = codebar_contracts::events::EventEnvelope;
pub type ErrorEnvelope = codebar_contracts::errors::ErrorEnvelope;
pub type ErrorCode = codebar_contracts::errors::ErrorCode;
pub type DomainResult<T> = Result<T, ErrorEnvelope>;

pub fn error_envelope(
    code: ErrorCode,
    message: impl Into<String>,
    retryable: bool,
) -> ErrorEnvelope {
    codebar_contracts::errors::ErrorEnvelope::new(code, message, retryable)
}

pub fn error_with_details(error: ErrorEnvelope, details: Value) -> ErrorEnvelope {
    codebar_contracts::errors::ErrorEnvelope::with_details(error, details)
}

File: packages/contracts/src/domain.ts
export type VcsType = "git" | "none";
export type TrustLevel = "trusted" | "untrusted";
export type ProviderKind = "claude" | "codex";
export type WorktreeSource = "existing" | "managed";
export type WorktreeLifecycleState = "preparing" | "ready" | "in_use" | "cleanup_pending" | "removed" | "error";
export type WorktreeCleanupPolicy = "manual" | "auto_on_task_done" | "keep";
export type TaskStatus = "draft" | "ready" | "active" | "blocked" | "completed" | "failed" | "cancelled" | "archived";
export type SessionLaunchMode = "new" | "resume";
export type SessionState = "draft" | "preparing_workspace" | "preparing_worktree" | "ready" | "launching" | "running" | "waiting_input" | "approval_required" | "interrupted" | "completed" | "failed" | "cancelled" | "archived";
export type LauncherType = "pty" | "headless";
export type RunExitReason = "completed" | "error" | "killed" | "crash" | "unknown";
export type RunAttemptStatus = "created" | "running" | "exited";
export type PlanMode = "guided" | "open";
export type PlanStatus = "draft" | "active" | "completed" | "cancelled";
export type PlanStepStatus = "pending" | "claimed" | "running" | "blocked" | "completed" | "cancelled";
export type SkillProfileSource = "workspace" | "worktree" | "task" | "step";
export type ApprovalActionType = "write" | "delete" | "git_push" | "dangerous_bash" | "external_side_effect";

export interface Workspace {
  id: string;
  displayName: string;
  rootPath: string;
  vcsType: VcsType;
  repoIdentity?: string;
  trustLevel: TrustLevel;
  defaultProvider?: ProviderKind;
  createdAt: string;
  updatedAt: string;
}

export interface Worktree {
  id: string;
  workspaceId: string;
  path: string;
  branchName?: string;
  baseBranch?: string;
  source: WorktreeSource;
  lifecycleState: WorktreeLifecycleState;
  cleanupPolicy: WorktreeCleanupPolicy;
  createdAt: string;
  updatedAt: string;
}

export interface Task {
  id: string;
  workspaceId: string;
  title: string;
  prompt: string;
  goal?: string;
  constraints?: string[];
  requestedProvider?: ProviderKind;
  requestedModel?: string;
  status: TaskStatus;
  activePlanId?: string;
  activeSkillProfileId?: string;
  createdAt: string;
  updatedAt: string;
}

export interface Session {
  id: string;
  taskId: string;
  workspaceId: string;
  worktreeId?: string;
  provider: ProviderKind;
  providerSessionId?: string;
  launchMode: SessionLaunchMode;
  state: SessionState;
  currentStepId?: string;
  lastActivityAt?: string;
  createdAt: string;
  updatedAt: string;
}

export interface RunAttempt {
  id: string;
  sessionId: string;
  attemptNo: number;
  launcherType: LauncherType;
  command: string;
  args: string[];
  cwd: string;
  pid?: number;
  startedAt?: string;
  endedAt?: string;
  exitReason?: RunExitReason;
  status: RunAttemptStatus;
}

export interface Plan {
  id: string;
  taskId: string;
  mode: PlanMode;
  status: PlanStatus;
  createdAt: string;
  updatedAt: string;
}

export interface PlanStep {
  id: string;
  planId: string;
  title: string;
  description?: string;
  status: PlanStepStatus;
  dependsOn: string[];
  parallelizable: boolean;
  requiredSkills: string[];
  allowedProviders?: ProviderKind[];
  leaseOwnerSessionId?: string;
  leaseToken?: string;
  leaseExpiresAt?: string;
  progressSummary?: string;
  progressDetails?: Record<string, unknown>;
  outputs?: Record<string, unknown>;
  blockedReason?: string;
  createdAt: string;
  updatedAt: string;
}

export interface SkillProfile {
  id: string;
  name: string;
  source: SkillProfileSource;
  workspaceId?: string;
  worktreeId?: string;
  taskId?: string;
  stepId?: string;
  allowedSkills: string[];
  preferredSkills?: string[];
  forbiddenSkills?: string[];
  createdAt: string;
  updatedAt: string;
}
